### PR TITLE
err: fix crash in ERR_load_strings() when configured with no-err

### LINKED ISSUE #17971: version 1.1.1L crashes with null pointer access in ERR_load_strings()
Hi all,
version: 1.1.1L
Built with option: no-err
Callstack:
pthread_rwlock_trywrlock(NULL)
pthread_rwlock_timedwrlock(NULL)
CRYPTO_THREAD_write_lock(NULL)
ERR_load_strings()
xmlSecOpenSSLInit()

Original openssl code:
```
int ERR_load_ERR_strings(void)
{
**#ifndef OPENSSL_NO_ERR**
    if (!RUN_ONCE(&err_string_init, do_err_strings_init))
        return 0;

    err_load_strings(ERR_str_libraries);
    err_load_strings(ERR_str_reasons);
    err_patch(ERR_LIB_SYS, ERR_str_functs);
    err_load_strings(ERR_str_functs);
    build_SYS_str_reasons();
#endif
    return 1;
}
```

If changed to:
```
int ERR_load_ERR_strings(void)
{
    if (!RUN_ONCE(&err_string_init, do_err_strings_init))
        return 0;

**#ifndef OPENSSL_NO_ERR**
    err_load_strings(ERR_str_libraries);
    err_load_strings(ERR_str_reasons);
    err_patch(ERR_LIB_SYS, ERR_str_functs);
    err_load_strings(ERR_str_functs);
    build_SYS_str_reasons();
#endif
    return 1;
}
```
Crash does not happen. Is this modification a valid fix?

Thanks,
Yusshi

### PR DESCRIPTION
**Pull request by mspncp**
Fixes #17971

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by mspncp** [comment]
_WIP: Waiting for approval of #17975 ..._

**Comment by mspncp** [comment]
This pull request should now be back in sync with #17975 again and is ready for review.

**Comment by mspncp** [comment]
note: the commits of this pull request can be rebased without conflicts onto ` openssl-3.0`

**Review by t8m — APPROVED** [review]
LGTM, but I'd like @levitte to also look.

File: crypto/err/err.c
Comment (levitte) [inline crypto/err/err.c:779]: This change is not ok.  The error library number isn't just about the strings, it also affects the error code itself, and is used to distinguish where an error comes from (extractable with `ERR_LIB`)

Comment (levitte) [reply crypto/err/err.c:779]: For example, every engine usually get themselves a unique error library number, to be able to report their own errors.  With this change, it's not possible which engine a particular engine generated error comes from, and since they can have overlapping reason codes, it's therefore not even possible to distinguish what the error was

Comment (levitte) [reply crypto/err/err.c:779]: The same goes for providers, btw...  See this:

https://github.com/openssl/openssl/blob/927d0566ded0dff9d6c5abc8a40bb84068446b76/crypto/provider_core.c#L555

**Review by levitte — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
As this PR currently stands, -1

File: crypto/err/err.c
Comment (mspncp) [reply crypto/err/err.c:779]: > This change is not ok. The error library number isn't just about the strings, it also affects the error code itself, and is used to distinguish where an error comes from (extractable with `ERR_LIB`)

Ah, I already pondered this question in https://github.com/openssl/openssl/pull/17975#issuecomment-1082315309. But a (probably too) quick search in 1.1.1 didn't bring up any indication that the number was used other than looking up the string. And @bernd-edlinger agreed that it was ok. 

> As this PR currently stands, -1

No problem, it's out of question that this part of my change needs to be reverted (in both PRs).

Comment (mspncp) [reply crypto/err/err.c:779]: @levitte may I ask you that you add this as a formal review change request to #17975 as well, so I can re-request review later? (One line of comment sufficies.)

**Comment by mspncp** [comment]
_(Actually, a -1 is not necessary, since I consider an unresolved change request as a blocking, even if the approval quota has been reached)_

File: crypto/err/err.c
Comment (bernd-edlinger) [reply crypto/err/err.c:779]: Hmm right, but you have no way to know which library ERR_LIB number refers to,
therefore I did not see what the number is good for, except tell you that a ERR_LIB_USER
caused the error.
Only the builtin numbers < 128 have fixed meaning, everything >= 128 just means,
you can only guess which module caused it, because the string table is missing.
And moreover, if ERR_get_next_error_library() is called more than 128 times, the
assigned ERR_LIB for those silently collide with the built-in ERR_LIB_XX defines.

Comment (mspncp) [reply crypto/err/err.c:779]: The library/provider could compare the number with the one it registered to distinguish its own error by the ones from other libraries/providers. 🤔

Comment (mspncp) [reply crypto/err/err.c:779]: > And moreover, if ERR_get_next_error_library() is called more than 128 times, the
> assigned ERR_LIB for those silently collide with the built-in ERR_LIB_XX defines.

This is certainly an error which should be catched (in another PR).

Comment (bernd-edlinger) [reply crypto/err/err.c:779]: Yes, but the only problem I see, is that no one of the current users of
ERR_get_next_error_library() has any idea how to do a proper error handling,
when this function returns 0.

Comment (bernd-edlinger) [reply crypto/err/err.c:779]: > The library/provider could compare the number with the one it registered to distinguish its own errors by the ones from other libraries/providers.

Yes, in theory, but since the uniqueness is not guaranteed, anything that depends on
the uniqueness is pretty fragile.

Comment (levitte) [reply crypto/err/err.c:779]: It's true that it's hard to find out what the error library number for an engine or provider is hard to get.  Than I'd say that's a bug / missing feature in itself (it shouldn't be that hard to add an accessor function for it), *not* a reason to obliterate the library numbers even more.

Comment (levitte) [reply crypto/err/err.c:779]: > @levitte may I ask you that you add this as a formal review change request to #17975 as well, so I can re-request review later? (One line of comment sufficies.)

Done

Comment (levitte) [reply crypto/err/err.c:779]: The 128 limit has been around for years, and has gone unnoticed as far as I know.  This is probably because people load only so many engines / providers at the same time (or have only so many OpenSSL compatible "libraries" in their application, which is also a possibility)...

Still, it's a valid question what should happen when that limit is reached!
I'd say that this is worth raising a separate issue.

Comment (bernd-edlinger) [reply crypto/err/err.c:779]: Hmm, in crypto/provider_core.c the function ERR_get_next_error_library();
is invoked here
https://github.com/openssl/openssl/blob/e4cdcb8bc44250aa4e0893dc4a7d64668f0fb949/crypto/provider_core.c#L1236
but it's result is never used here
https://github.com/openssl/openssl/blob/e4cdcb8bc44250aa4e0893dc4a7d64668f0fb949/crypto/provider_core.c#L1906-L1910
because ERR_GET_LIB(reason) is never zero, but instead most of the time ERR_LIB_PROV
but the reason code may be from a different openssl version.

**Comment by mspncp** [comment]
I just dropped the last commit 5b0f6b8655a1f56f3ac870ed283473e7078ab439.

**Comment by mspncp** [comment]
ping

**Review by t8m — APPROVED** [review]

**Comment by mspncp** [comment]
@paulidale since you approved #17975, would you mind approving this one as well?

**Review by paulidale — APPROVED** [review]

**Comment by tmshort** [comment]
Are we still waiting on @levitte ?

**Comment by t8m** [comment]
IMO this should be merged as this is basically the same fix as on 1.1.1 @mspncp ?

**Comment by mspncp** [comment]
Thanks for the reminder, merged to master and openssl-3.0:

**master**
* 1c8787d5e0 err: get rid of err_free_strings_int()
* 11e85b8941 err: fix crash in ERR_load_strings() when configured with no-err

**openssl-3.0**
* 39f7f54373 err: get rid of err_free_strings_int()
* 3e8307924e err: fix crash in ERR_load_strings() when configured with no-err

**Pull request closed without merging**
